File: Java/144.java
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        LinkedList<TreeNode> stack = new LinkedList<>();
        LinkedList<Integer> output = new LinkedList<>();

        if (root == null) {
            return output;
        }

        stack.add(root);

        while(!stack.isEmpty()){
            TreeNode tree = stack.pollLast();

            output.add(tree.val);

            if (tree.right != null) {
                stack.add(tree.right);
            }

            if (tree.left != null) {
                stack.add(tree.left);
            }

        }

        return output;

    }
}

File: JavaScript/144.js
/**
 * 
 */

// 递归
var preorderTraversal = function(root) {
    return root ? [root.val, ...preorderTraversal(root.left), ...preorderTraversal(root.right)] : [];
};

// 迭代
var preorderTraversal = function(root) {
    let arr = [];
    let res = [];    
    
    root && arr.push(root)
    
    while(arr.length > 0) {
        let cur = arr.pop()
        res.push(root.val)
        cur.right && arr.push(cur.right) 
        cur.left && arr.push(cur.left)
    }    
    return arr
}